### PR TITLE
feat: identify if a token is being approved and estimate gas for contract trxs

### PR DESCRIPTION
**Pull request by richard-ramos**

### CODE REVIEW

File: ui/app/AppLayouts/Chat/ChatColumn/ChatComponents/SignTransactionModal.qml [hunk 1]
```diff
@@ -87,6 +100,7 @@ ModalPopup {
                 }
 
                 function estimateGas() {
+                    console.log("CALLING ESTIMATE GAS")
```
Comment (jrainville) [inline ui/app/AppLayouts/Chat/ChatColumn/ChatComponents/SignTransactionModal.qml:103]:
```suggestion
```

**Review by jrainville — APPROVED** [review]

File: ui/app/AppLayouts/Chat/ChatColumn/ChatComponents/SignTransactionModal.qml [hunk 2]
```diff
@@ -68,7 +74,14 @@ ModalPopup {
...
             //% "Send"
-            headerText: qsTrId("command-button-send")
+            headerText: {
+                if(trxData.startsWith("0x095ea7b3")){
+                    const approveData = JSON.parse(walletModel.decodeTokenApproval(selectedRecipient.address, trxData))
+                    if(approveData.symbol)
+                        return qsTr("Authorize %1 %2").arg(approveData.amount).arg(approveData.symbol)    
+                }
+                return qsTrId("command-button-send");
+            }
```
Comment (emizzle) [inline ui/app/AppLayouts/Chat/ChatColumn/ChatComponents/SignTransactionModal.qml:84]: Where is the encoded data coming from? IOW, why is the gas not already estimated at the origin?

**Review by emizzle — CHANGES_REQUESTED** [review]
Very cool. I'm curious where the encoded data comes from in the first place? 

Will the estimation of gas be needed for other methods other than just token approval? In that case, we'd likely need to make this more generic:

1. "Read" the ABI-encoded method from the first bytes of the data
2. From the method name, find the associated strongly-typed nim object, ie `type Approval* = object...`
3. decode the rest of the data in to the nim object `let decoded = decode[Approve](data)`

Leaving as "request changes" due to the divergence from master as noted in the comments.

File: ui/app/AppLayouts/Chat/ChatColumn/ChatComponents/SignTransactionModal.qml (hunk 2 above)
Comment (richard-ramos) [reply ui/app/AppLayouts/Chat/ChatColumn/ChatComponents/SignTransactionModal.qml:84]: The encoded data comes from any transaction created in the browser. Having a gas estimation will depend on whether the dapp developer did it or not. so.. :thinking:  we probably should check if the payload contains a gasLimit before calling estimateGas ourselves

**Pull request merged**